File: src/keywords/keyword_loader.py
import csv
from pathlib import Path
from typing import Dict, List

KEYWORD_FILE = Path(__file__).resolve().parents[2] / "data" / "IT_Security_Keywords.csv"


def normalize_keyword(raw_value: str) -> str:
    """Normalize a keyword by removing numeric prefixes and trimming whitespace."""
    term = raw_value.strip().lower()
    # Remove leading numeric column prefixes like '00.' or '1.'
    while term and term[0].isdigit():
        term = term[1:]
    term = term.lstrip('. ')  # strip separators after the prefix
    return term


def normalize_category(raw_value: str) -> str:
    """Normalize a category label by removing numeric prefixes and trimming whitespace."""
    return normalize_keyword(raw_value).title()


def load_keyword_categories() -> Dict[str, List[str]]:
    """Load CSV columns as category -> keywords using normalized first-row headers."""
    if not KEYWORD_FILE.exists():
        return {}

    with KEYWORD_FILE.open(newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile, delimiter=';')
        rows = list(reader)

    if not rows:
        return {}

    categories = [normalize_category(cell) for cell in rows[0]]
    keyword_map: Dict[str, List[str]] = {
        category: []
        for category in categories
        if category
    }
    seen_by_category = {category: set() for category in keyword_map}

    for row in rows[1:]:
        for index, cell in enumerate(row):
            if index >= len(categories):
                continue

            category = categories[index]
            if not category:
                continue

            term = normalize_keyword(cell)
            if term and term not in seen_by_category[category]:
                seen_by_category[category].add(term)
                keyword_map[category].append(term)

    return {category: keywords for category, keywords in keyword_map.items() if keywords}


def load_keywords() -> List[str]:
    """Load all unique keywords from the categorized CSV file."""
    keywords = []
    seen = set()

    for category_keywords in load_keyword_categories().values():
        for term in category_keywords:
            if term and term not in seen:
                seen.add(term)
                keywords.append(term)

    return keywords


def get_keyword_examples(limit: int = 10) -> List[str]:
    """Return a small set of example keywords."""
    keywords = load_keywords()
    return keywords[:limit] if keywords else []

File: php/src/KeywordLoader.php
<?php

namespace Ibc;

/**
 * Port of src/keywords/keyword_loader.py
 */
class KeywordLoader
{
    public static function keywordFile(): string
    {
        return dirname(__DIR__) . '/data/IT_Security_Keywords.csv';
    }

    public static function normalizeKeyword(string $rawValue): string
    {
        $term = strtolower(trim($rawValue));

        while ($term !== '' && ctype_digit($term[0])) {
            $term = substr($term, 1);
        }

        return ltrim($term, ". ");
    }

    public static function normalizeCategory(string $rawValue): string
    {
        $term = self::normalizeKeyword($rawValue);

        return ucwords($term, " \t\r\n\f\v-'");
    }

    /**
     * @return array<string, string[]>
     */
    public static function loadKeywordCategories(): array
    {
        $file = self::keywordFile();

        if (!is_file($file)) {
            return [];
        }

        $rows = [];
        $handle = fopen($file, 'r');
        if ($handle === false) {
            return [];
        }

        while (($row = fgetcsv($handle, 0, ';')) !== false) {
            $rows[] = $row;
        }
        fclose($handle);

        if (empty($rows)) {
            return [];
        }

        $categories = array_map(
            fn ($cell) => self::normalizeCategory((string) $cell),
            $rows[0]
        );

        $keywordMap = [];
        $seenByCategory = [];
        foreach ($categories as $category) {
            if ($category !== '') {
                $keywordMap[$category] = [];
                $seenByCategory[$category] = [];
            }
        }

        for ($r = 1; $r < count($rows); $r++) {
            $row = $rows[$r];
            foreach ($row as $index => $cell) {
                if ($index >= count($categories)) {
                    continue;
                }

                $category = $categories[$index];
                if ($category === '') {
                    continue;
                }

                $term = self::normalizeKeyword((string) $cell);
                if ($term !== '' && !isset($seenByCategory[$category][$term])) {
                    $seenByCategory[$category][$term] = true;
                    $keywordMap[$category][] = $term;
                }
            }
        }

        return array_filter($keywordMap, fn ($keywords) => !empty($keywords));
    }

    /**
     * @return string[]
     */
    public static function loadKeywords(): array
    {
        $keywords = [];
        $seen = [];

        foreach (self::loadKeywordCategories() as $categoryKeywords) {
            foreach ($categoryKeywords as $term) {
                if ($term !== '' && !isset($seen[$term])) {
                    $seen[$term] = true;
                    $keywords[] = $term;
                }
            }
        }

        return $keywords;
    }
}
